compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3743,6 +3881,12 @@
             // Handle other types except for semantic errors
             resultType = ((BFutureType) resultType).constraint;
         }
+
+        if (waitExpr.getExpression().getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
```

Comment:
regular wait is when size of getExpression is one.
We need to support non simple_var_ref situations.

```ballerina
// Coppied from other comment
function bar() returns future<int> {
    function () returns int f = () => 0;
    return start f();
}

public function main() {
    int x = wait bar();
}
```